src/amplify-migration/jsDomUtils.ts

```diff
@@ -0,0 +1,186 @@
+
+      if (changedPermalinks[rawPermalink]) {
+        img[attribute] = img.src.replace(
+          rawPermalink,
+          changedPermalinks[rawPermalink]
+        );
+      }
+
+      const { fileContent:src } =
+        await updateFilesUploadsPath(
+          img.src,
+          setOfAllDocumentsPath,
+          currentRepoName
+        );
+      if (img.src !== src) {
+        fileContent.replace(img.src, src)
+      }
+    }
+  }
+  return { fileContent, dom, changedPermalinks };
+}
+/**
+ * Requirements:
+ * convert `/some/path` to `some/path`
+ * convert `some/path/` to `some/path`
+ * above two variants with `permalink: ` prefix
+ * above two variants with `url: ` prefix
+ * @param permalink original permalink
+ * @returns raw permalinks without leading/trailing slash
+ */
+
+export function getRawPermalink(permalink: string) {
+  let trimmedPermalink = permalink.trim();
+  if (trimmedPermalink.startsWith(`permalink: `)) {
+    trimmedPermalink = permalink.trim().slice(11);
+  }
+  if (trimmedPermalink.startsWith(`url: `)) {
+    trimmedPermalink = permalink.trim().slice(5);
+  }
+  if (trimmedPermalink.startsWith(`/`)) {
+    trimmedPermalink = trimmedPermalink.slice(1);
+  }
+  if (trimmedPermalink.endsWith("/")) {
+    trimmedPermalink = trimmedPermalink.slice(0, -1);
+  }
+
+  return trimmedPermalink;
+}
+/**
+ * We need to modify trailing slashes from files. We get the list of added extensions from
+ * https://github.com/isomerpages/isomercms-backend/blob/develop/src/utils/file-upload-utils.js
+ */
+
+export async function updateFilesUploadsPath(
+  fileContent: string,
+  setOfAllDocumentsPath: Set<string>,
+  currentRepoName: string
+): Promise<{ fileContent: string }> {
+  const fileRegexWithTrailingSlash = /\/(files|images)\/.*.(pdf|png|jpg|gif|tif|bmp|ico|svg)\//g;
+  const matches = fileContent.match(fileRegexWithTrailingSlash);
+  if (matches) {
+    matches.forEach(async (match) => {
+      assert(match.endsWith("/")); // sanity check that should have been guaranteed by regex
+      let newFilePath = match.slice(0, -1);
+      fileContent = fileContent.replace(match, newFilePath);
+      if (!newFilePath.startsWith("/")) {
+        // this is needed since setOfAllDocumentsPath has leading slash
+        newFilePath = "/" + newFilePath;
+      }
+      
+    });
+  }
+  // check that files actually exist, else change file names (not folders)
+  const fileRegex = /\/(files|images)\/.*.(pdf|png|jpg|gif|tif|bmp|ico|svg)/g;
+  const fileMatches = fileContent.match(fileRegex);
+  if (fileMatches) {
+    fileMatches.forEach(async (match) => {
+      let doesFileExist = false
+      for (const path of setOfAllDocumentsPath) {
+        if (path === match || decodeURIComponent(match) === path) {
+          doesFileExist = true;
+          break;
+        }
+      }
+      if (!doesFileExist) {
+        // see if we can self-recover from this by modifying the permalink directly
```

Comment:
During migration can we actually run through all existing folders and files and perform a lowercase operation on it?

Since we are already editing the permalinks/file links, can just .toLowerCase() for all of them and replace in the file content?

Reply:
Ah re-read your PR on the alternative solution section.Regarding the concerns,

1. How likely do you think this is? First thought is that people won't have /EXAMPLE/test.md and /example/test.md simultaneously
2. Why do we worry about special chars here, we are just making everything lowercase and keep the rest as it is right?

Rather than conforming to case sensitivity, making things case insensitive seems more intuitive to me

Reply:
> How likely do you think this is? First thought is that people won't have /EXAMPLE/test.md and /example/test.md simultaneously

Originally I was worried about the possibility of this since these scripts will be running on repos that used to exist even before the CMS existed bah

Reply:
It kinda depends on what we are optimising for here, it could be: 

1. Minimal code changes? (current implementation)
2. Backward compat? ie links that already exist in the md should continue to exist? (Hardest to implement)
3. Consistency for all files? (Suggested implementation)